File: aliasengine/aliasengine/Program/Index/UpdateProgram.swift
//
//  UpdateProgram.swift
//  aliasengine
//
//  Created by Gabriel Coman on 10/06/2022.
//

import Foundation

class UpdateProgram: Program {
    private let reader = DataReader()
    private let writer = DataWriter()
    
    func run() {
        let sources = reader.readSources()
        var items: [IndexItem] = []
        
        for source in sources {
            switch source.type {
            case .alias:
                items += reloadAliasFiles(source: source)
            case .script:
                items += reloadScriptFiles(source: source)
            }
        }
        
        writer.write(items: items.unique { $0.name })
    }
    
    private func reloadAliasFiles(source: SourceFile) -> [IndexItem] {
        guard let file = try? String(contentsOfFile: source.path) else {
            print("Could not open file \(source.path, color: .red)")
            return []
        }
        
        let aliasIngester = AliasIngester(filePath: source.path)
        let newAliases = aliasIngester.process(fileContents: file)
        
        let functionIngester = FunctionIngester(filePath: source.path)
        let newFunctions = functionIngester.process(fileContents: file)
        
        print("Ingested \(newAliases.count + newFunctions.count) aliases and functions from \(source.path)")
        
        return (newAliases + newFunctions)
    }
    
    private func reloadScriptFiles(source: SourceFile) -> [IndexItem] {
        guard let content = try? String(contentsOfFile: source.path) else {
            print("Could not open file \(source.path, color: .red)")
            return []
        }
        
        let ingester = ScriptIngester(withAlias: source.name, andDiskPath: source.path)
        
        print("Ingested file \(source.name) at \(source.path)")
        
        return ingester.process(fileContents: content)
    }
}

File: bee/update_program.go
package main

import (
	"bee/bbee/data"
	"bee/bbee/models"
	"fmt"
)

type UpdateProgram struct {
}

func (u UpdateProgram) run() {
	var result = []models.IndexItem{}
	var sources = data.ReadSources()

	for _, source := range sources {
		switch source.Type {
		case models.SourceType(models.Command):
			items := u.updateConfigFiles(source)
			result = append(result, items...)
		case models.SourceType(models.File):
			items := u.updateScriptFiles(source)
			result = append(result, items...)
		}
	}

	result = uniqueItemsByDate(result)
	data.WriteItems(result)

	fmt.Printf("Updated %d elements\n", len(result))
}

func (u UpdateProgram) updateConfigFiles(source models.SourceFile) []models.IndexItem {
	// open file
	contents, err := data.ReadFile(source.Path)

	// gently handle error
	if err != nil {
		fmt.Println(err)
		return []models.IndexItem{}
	}

	time := CurrentTime()
	ingester := ConfigIngester{filePath: source.Path, currentTime: time}
	return ingester.process(contents)
}

func (u UpdateProgram) updateScriptFiles(source models.SourceFile) []models.IndexItem {
	// open file
	contents, err := data.ReadFile(source.Path)

	// gently handle error
	if err != nil {
		fmt.Println(err)
		return []models.IndexItem{}
	}

	time := CurrentTime()
	ingester := ScriptIngester{alias: source.Name, path: source.Path, currentTime: time}
	return ingester.process(contents)
}
